From 96d3bb55ea82227ce2962c4a8afe8024fa9380a0 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Mon, 2 Dec 2024 15:47:16 +0100
Subject: [PATCH 1389/1695] clk: stm32mp21: do not configure PLL1 MUX when
 M33TDCID

Do not access PLL mux when M33 is TDCID. Just check if the mux is correct.

Alignment with STM32MP25 driver patch reviewed on
https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/411803

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git ecc5d4f4b917376b7d6b5dcfe4442d2d5e85c365 clk: stm32mp21: do not configure PLL1 MUX when M33TDCID]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: Iaebc3718c2456ecf93c51a63c7b4cf103142559f
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/422150
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/drivers/clk/clk-stm32mp21.c | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/core/drivers/clk/clk-stm32mp21.c b/core/drivers/clk/clk-stm32mp21.c
index 50f9af650..be77b91ac 100644
--- a/core/drivers/clk/clk-stm32mp21.c
+++ b/core/drivers/clk/clk-stm32mp21.c
@@ -1755,6 +1755,15 @@ static int clk_stm32_pll_set_mux(struct clk_stm32_priv *priv __unused,
 	return 0;
 }
 
+static int clk_stm32_pll_check_mux(struct clk_stm32_priv *priv __unused,
+				   uint32_t src)
+{
+	int mux = (src & MUX_ID_MASK) >> MUX_ID_SHIFT;
+	int sel = (src & MUX_SEL_MASK) >> MUX_SEL_SHIFT;
+
+	return stm32_mux_get_parent(mux) != (size_t)sel;
+}
+
 static int clk_stm32_pll1_init(struct clk_stm32_priv *priv,
 			       int pll_idx __unused,
 			       struct stm32_pll_dt_cfg *pll_conf)
@@ -1770,7 +1779,11 @@ static int clk_stm32_pll1_init(struct clk_stm32_priv *priv,
 
 	stm32mp2_a35_ss_on_bypass();
 
-	ret = clk_stm32_pll_set_mux(priv, pll_conf->src);
+	if (stm32_rcc_has_access_by_id(RCC_RIF_PLL4_TO_8))
+		ret = clk_stm32_pll_set_mux(priv, pll_conf->src);
+	else
+		ret = clk_stm32_pll_check_mux(priv, pll_conf->src);
+
 	if (ret != 0)
 		panic();
 
-- 
2.47.2

